packages/grpc-js/src/server.ts

```diff
@@ -724,150 +733,233 @@ export class Server {
-              break;
-            default:
-              throw new Error(`Unknown handler type: ${handler.type}`);
-          }
-        } catch (err) {
-          if (!call) {
-            call = new Http2ServerCallStream(stream, null!, this.options);
-            if (this.channelzEnabled) {
-              this.callTracker.addCallFailed();
-              channelzSessionInfo?.streamTracker.addCallFailed()
-            }
-          }
+    if (handler === undefined) {
+      this.trace(
+        'No handler registered for method ' +
+        path +
+        '. Sending UNIMPLEMENTED status.'
+      );
+      throw getUnimplementedStatusResponse(path);
+    }
 
-          if (err.code === undefined) {
-            err.code = Status.INTERNAL;
-          }
+    return handler
+  }
+  
+  private _respondWithError<T extends Partial<ServiceError>>(
+    err: T, 
+    stream: http2.ServerHttp2Stream, 
+    channelzSessionInfo: ChannelzSessionInfo | null = null
+  ) {
+    const call = new Http2ServerCallStream(stream, null!, this.options);
+    
+    if (err.code === undefined) {
+      err.code = Status.INTERNAL;
+    }
+
+    if (this.channelzEnabled) {
+      this.callTracker.addCallFailed();
+      channelzSessionInfo?.streamTracker.addCallFailed()
+    }
 
-          call.sendError(err);
+    call.sendError(err);
+  }
+
+  private _channelzHandler(stream: http2.ServerHttp2Stream, headers: http2.IncomingHttpHeaders) {
+    const channelzSessionInfo = this.sessions.get(stream.session as http2.ServerHttp2Session);
+    
+    this.callTracker.addCallStarted();
+    channelzSessionInfo?.streamTracker.addCallStarted();
+
+    if (!this._verifyContentType(stream, headers)) {
+      this.callTracker.addCallFailed();
+      channelzSessionInfo?.streamTracker.addCallFailed();
+      return
+    }
+
+    let handler: Handler<any, any>
+    try {
+      handler = this._retrieveHandler(headers)
+    } catch (err) {
+      this._respondWithError(err, stream, channelzSessionInfo)
+      return
+    }
+  
+    const call = new Http2ServerCallStream(stream, handler, this.options);
+      
+    call.once('callEnd', (code: Status) => {
+      if (code === Status.OK) {
+        this.callTracker.addCallSucceeded();
+      } else {
+        this.callTracker.addCallFailed();
+      }
+    });
+    
+    if (channelzSessionInfo) {
+      call.once('streamEnd', (success: boolean) => {
+        if (success) {
+          channelzSessionInfo.streamTracker.addCallSucceeded();
+        } else {
+          channelzSessionInfo.streamTracker.addCallFailed();
         }
+      });
+      call.on('sendMessage', () => {
+        channelzSessionInfo.messagesSent += 1;
+        channelzSessionInfo.lastMessageSentTimestamp = new Date();
+      });
+      call.on('receiveMessage', () => {
+        channelzSessionInfo.messagesReceived += 1;
+        channelzSessionInfo.lastMessageReceivedTimestamp = new Date();
+      });
+    }
+
+    const metadata = call.receiveMetadata(headers);
```

Comment:
As far as I can tell, starting at this point this function is the same as `_streamHandler`, except for the last branch of the `if` statements. I would prefer that this section of code be factored into a single function that both handler functions use. That last branch is not expected to be reachable in normal operation so I think it would be reasonable for that new function to conditionally perform the channelz operations in that branch at call handling time.